src/utils/scrap_data.ts

```diff
@@ -0,0 +1,176 @@
+import puppeteer, { Browser, Page } from 'puppeteer';
+import { VIVINO_GRAPES_API, VIVINO_SEARCH_URL, VIVINO_EXPLORE_API, WINES_PER_PAGE, VIVINO_WINE_STYLES_API, VIVINO_FOODS_API } from '../const/const';
+import { REGIONS } from '../const/regions';
+import { WINE_TYPES } from '../const/wine_types';
+import { WineParams } from '../interfaces/params.interface';
+import { EXPLORE_API_PARAMS } from '../enums/explore_api_params.enum';
+import { Grape, Wine, WineStyle, Food } from '../interfaces/wine.interface';
+
+class ScrapData {
+  grapes: Grape[] = [];
+  wineStyles: WineStyle[] = [];
+  foods: Food[];
+
+  appendParam = (param: string, queryKey: string, list: any = []) => {
+    if (!param) {
+      return '';
+    }
+    let queryString = '';
+    const paramNames = param.split(',');
+    paramNames.forEach(name => {
+      if (list.length == 0) {
+        queryString += `&${queryKey}=${name}`;
+      } else {
+        const obj = list.find((g: any) => g.name.toUpperCase().indexOf(name.toUpperCase()) > -1);
+        if (obj) {
+          queryString += `&${queryKey}=${obj.id}`;
+        }
+      }
+    });
+    return queryString;
+  };
+
+  getVivinoSearchUrl = (pageIndex: number, wineParams: WineParams): string => {
+    wineParams.minPrice = isNaN(wineParams.minPrice) ? 0 : wineParams.minPrice;
+    wineParams.maxPrice = isNaN(wineParams.maxPrice) ? 99999999 : wineParams.maxPrice;
+    wineParams.minRating = isNaN(wineParams.minRating) ? 1 : wineParams.minRating;
+    wineParams.orderBy = !wineParams.orderBy ? 'ratings_average' : wineParams.orderBy;
+
+    let url = `${VIVINO_EXPLORE_API}?${EXPLORE_API_PARAMS.country_code}=US&${EXPLORE_API_PARAMS.currency_code}=USD&${EXPLORE_API_PARAMS.grape_filter}=varietal`;
+    url += `&${EXPLORE_API_PARAMS.order_by}=${wineParams.orderBy}&${EXPLORE_API_PARAMS.order}=desc&${EXPLORE_API_PARAMS.page}=${pageIndex}`;
+    url += this.appendParam(wineParams.grapes, EXPLORE_API_PARAMS.grape_ids, this.grapes);
+    url += this.appendParam(wineParams.wineStyles, EXPLORE_API_PARAMS.wine_style_ids, this.wineStyles);
+    url += this.appendParam(wineParams.foodPairings, EXPLORE_API_PARAMS.food_ids, this.foods);
+    url += `&${EXPLORE_API_PARAMS.min_rating}=${wineParams.minRating}`;
+    url += `&${EXPLORE_API_PARAMS.price_range_min}=${wineParams.minPrice}`;
+    url += `&${EXPLORE_API_PARAMS.price_range_max}=${wineParams.maxPrice}`;
+    if (wineParams.countryCodes) {
```

Comment:
When I ran your app I got an error (just 0 items as I remember) when I passed countryCodes.
Why it is so and how you would fix it?
If it's not possible to fix then why do you save the ability to pass this parameter?

Reply:
Actually, I could get the response when I passed the countryCodes.
`ts-node src/index.ts --grapes="malbec,merlot" --minRating=1 --minPrice=30 --maxPrice=40 --minNoRatings=10 --maxNoRatings=10000 --foodPairings="Poultry" --wineTypes="red,white" --countryCodes="us"`
This is the command I used to run.

But if the country code is not valid, I am getting an error. I noticed that error only now.
This error can be avoided by implementing country code validation.

Reply:
Could you try to run for Canada, please?

Reply:
I ran for Canada with the following command.
`ts-node src/index.ts --grapes="malbec,merlot" --minRating=1 --minPrice=30 --maxPrice=40 --minNoRatings=10 --maxNoRatings=10000 --foodPairings="Poultry" --wineTypes="red,white" --countryCodes="ca"`

I got the response from vivino API as follows
```
{
  market: {
    country: 'us',
    state: 'California',
    currency: { code: 'USD', name: 'US Dollars', prefix: '$', suffix: null }
  },
  records_matched: 0,
  matches: [],
  bottle_type_errors: [],
  records: []
}
```

I checked in the website also and no result is returning for the given params.
![Screenshot 2020-12-03 at 4 22 43 PM](https://user-images.githubusercontent.com/68052759/101000348-c8c0d280-3583-11eb-957f-49944cff0ab3.png)

Reply:
I ran the following command and could get the result with Canada.
`ts-node src/index.ts --grapes="malbec,merlot" --minRating=1 --minPrice=0 --maxPrice=40 --foodPairings="Poultry" --wineTypes="red,white" --countryCodes="ca"`

Result
```
[
  {
    title: 'La Frenz Rockyfeller Vyd Malbec 2016',
    vintage: 2016,
    averagePrice: 24.73,
    currency: 'USD',
    averageRating: 4,
    numberOfRatings: 43,
    imageUrl: '//images.vivino.com/labels/UAUELcKYSgaSt8Fxz2LluQ.jpg',
    region: 'Naramata Bench',
    country: 'Canada',
    bottleVolume: 750,
    discountPercent: null
  }
]
```